subsys/bluetooth/audio/ascs.c

```diff
@@ -879,23 +872,22 @@ static void ascs_ep_iso_disconnected(struct bt_bap_ep *ep, uint8_t reason)
 
 	if (ep->status.state == BT_BAP_EP_STATE_RELEASING) {
 		ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);
-	} else if (ep->status.state == BT_BAP_EP_STATE_STREAMING) {
-		/* CIS has been unexpectedly disconnected */
-		ase->unexpected_iso_link_loss = true;
+	} else if (ep->status.state == BT_BAP_EP_STATE_STREAMING ||
+		   ep->status.state == BT_BAP_EP_STATE_DISABLING) {
+		int err;
 
-		/* The ASE state machine goes into different states from this operation
-		 * based on whether it is a source or a sink ASE.
+		/* ASCS_v1.0 3.2 ASE state machine transitions
+		 *
+		 * If the server detects link loss of a CIS for an ASE in the Streaming state or the
+		 * Disabling state, the server shall immediately transition that ASE to the QoS
+		 * Configured state.
 		 */
-		if (ep->dir == BT_AUDIO_DIR_SOURCE) {
-			ascs_ep_set_state(ep, BT_BAP_EP_STATE_DISABLING);
-		} else {
-			ascs_ep_set_state(ep, BT_BAP_EP_STATE_QOS_CONFIGURED);
-		}
-	} else if (ep->status.state == BT_BAP_EP_STATE_DISABLING) {
-		/* CIS has been unexpectedly disconnected */
-		ase->unexpected_iso_link_loss = true;
+		ase->state_pending = BT_BAP_EP_STATE_QOS_CONFIGURED;
 
-		ascs_ep_set_state(ep, BT_BAP_EP_STATE_QOS_CONFIGURED);
+		err = k_work_submit(&ase->state_transition_work);
+		if (err < 0) {
+			LOG_ERR("Failed to submit state transition work err %d", err);
+		}
```

Comment:
So I assume you are doing it this way, instead of `ascs_ep_set_state(ep, BT_BAP_EP_STATE_QOS_CONFIGURED);` because of the checks we have in `ascs_ep_set_state`?

What about we still set `ase->unexpected_iso_link_loss = true;` and add that as a check in `ascs_ep_set_state` to allow for `STREAMING->QOS_CONFIGURED` or `DISABLING->QOS_CONFIGURED` if `unexpected_iso_link_loss` is true?

Reply:
That's a good idea